아마존 캐러셀 HTML, CSS 구현

sangwon21 opened the pull request:
- [page 예시](https://sangwon21.github.io/amazon-carousel/slider.html)
- what
    - 아마존 슬라이더 이동을 구현했습니다.
    - 슬라이더 이동 이벤트를 등록했습니다.
    - 왼쪽 오른쪽 이동을 다르게 계산하는 알고리즘을 작성했습니다.

- how
    - ES6 Class 문법을 활용하여 UI Component를 객체로 나눴습니다.
    - ES6 Class 문법 중 상속을 활용하여 event객체가 갖는 공통의 기능을 부모 객체에 주었습니다.
    - 이동을 할 때, transitionend 이벤트를 활용하여 사용자를 속였습니다.

- why
    - 전체적으로 class를 활용하여 객체 간 관계를 설정하는 데에 집중하여 코드를 작성하고 싶었습니다. 이를 위해, 객체가 서로 의존성을 맺게 하거나, 상속을 통하여 관계를 지어줬습니다. 
    - 하지만 관계 설정을 정교하게 하지 못한 결과, event객체를 구성하면서 많은 어려움이 있었습니다. 실례로, SliderEvent의 특정 이벤트 발생 후, HeaderEvent 속 이벤트를 발생시켜야 하는데, 작성해둔 구조에서는 이벤트를 발생시키기 어려운 구조였습니다. 다른 방법으로, 우회해서 해결했지만, 어설픈 구조가 주는 어려움을 이번에 겪었습니다. 또한, 굳이 상속을 쓸 필요가 있었나라는 생각도 듭니다. 일단, 객체 간 공통 요소가 보여, 그 요소를 부모 요소로 빼는 작업을 진행했는데, 그 부모 요소의 규모가 작아서, 괜히 복잡성만 코드에 더한 게 아닐까 고민됩니다. 

- 질문 사항
    - bind를 해서 this context를 강제하는 게 과연 올바른 함수 사용법인지 궁금합니다.
    - 이벤트 전달을 고려하여 코드를 작성하다 보니, 재활용성이 많이 떨어지는 함수를 작성했습니다. 특히, HTML 구조가 바꼈을 때, 유연하게 대처할 수 있는, 이벤트 전달을 고려한 함수를 작성하기 위해서는, 어떠한 영역이 필요할까요?
    - card 객체의 sortByCategory() 매서드 호출하는 곳을 살펴보면, for문 한 번으로 대체할 수 있습니다. 대신 sortByCategory()함수는 조금 더 가독성이 좋은 것 같습니다. 이런 상황에서 어떤 선택을 해야 하는지 궁금합니다.

Commit:
[Reese, won] Slider UI HTML, CSS 작성 (#114)

* Complete basic structure

- Add util.js for the sake of convenience

* Complete header html rendering

* Complete card html rendering

* Complete button html rendering

* Complete header css

* Complete card css

* Complete button css

* Format Style

- Consistency in tab indent

* Fix type

sangwon21 committed:
feat: Header CSS 수정

why: 아마존 슬라이더와 비슷하게 CSS를 수정했다. 기존의 문구만 있던
부분에서, 배경 이미지를 준 CSS로 수정했다.

sangwon21 committed:
feat: 헤더 클릭 시 이벤트 추가

why: 헤더의 그림을 클릭하면, 해당 상자가 커져야 한다. 해당 상자 옆으로
그림자 효과도 추가되어야 한다. 마지막으로 해당 상자의 숨겨진 점들이
드러나야 한다. 이를 위해서 이벤트를 추가했는데, 이벤트 객체를 분리해서
온전히 이벤트 추가만 담당하게 만들었다. 하지만, 이벤트 버블링을 고려한
함수를 만들다 보니, 함수가 하는 일이 지나치게 커졌다. 이를 줄이기 위한
노력이 필요하다.

sangwon21 committed:
feat: 가짜 데이터 변경

why: 데이터 구조에 맞게 가짜 데어터를 변경했다. 프로그램에서 필요한
데이터를 추가했다. 이에 맞춰, 데이터를 다루기 편하게, 객체 형태로
저장했다.

sangwon21 committed:
feat: 이벤트 객체 생성 및 분리

why: 기존코드에서 slider-event에서 모든 이벤트의 책임을 지고 있었다.
하지만 헤더 영역과 슬라이더 영역에서 서로 필요한 이벤트의 성격이 다름에
따라 분리할 필요성이 생겼다. 코드를 작성하다 보니, 이벤트를 실제
이벤트가 적용되는 영역에 붙여주는 코드가 중복되는 것을 발견했다. 이
부분은 상속을 통해 다른 하나의 객체로 분리해 주었다.

sangwon21 committed:
feat: 슬라이더 기능 구현

why:
- 버튼의 next나 prev를 클릭하면 카드가 다음 방향이나 이전 방향으로 움직이는 기능을 구현했다. 개발할 때, 편의상 overflow: hidden값을 주지 않았다. 카드 1장당 id를 1,2,3... 이런 식으로 정수로 주고, css를 이런정수에 적용시켜 보려고 했다. 하지만 css 파일에서 정수를 인식하지 못하여, 대신 card + 정수 이런 id값을 활용하였다.
- 슬라이더 구현 방식은 다음과 같다. 카메라는 긴 띠의 중앙만 항상 촬영한다. 그래서 미리 띠를 중앙으로 옮겨야한다. 이 때, 옮기는 방식은 다양한데, transform만큼은 활용하면 안 된다. 왜냐하면 후에 상술할, 카드 이동방식이랑 css가 겹쳐, 원하는 효과가 나타나지 않는다. 여기서는, left에-값을 주어, 중앙에 배치되게끔 했다. 사용자가 버튼을 클릭하면, 해당 방향에 맞게 띠를 한 칸 이동해준다. transition이 끝나면 옮긴 칸이 중앙에 오게끔 배열을 조작한다. 이 때, transition 효과를 none으로 주어, 어떠한 움직임도 화면에 나타나지 않게한다. 배열 조작 후, translateX(0)으로 주어 제자리에 있게끔 한다.

sangwon21 committed:
feat: 헤더의 점 클릭시 선택 표시

why: 헤더를 클릭했을 때, 점을 표시해준다. 이벤트가 다양한 영역에서
발생할 수 있기 때문에, if문을 통하여 각각의 경우를 처리했다. 또한, dot을
정확히 클릭했을 때, dot이 사라지는 경우가 발생했다. 이는 invisible
클래스를 잘못된 tag에 줘서 발생항 문제로, 다시 제대로 된 tag에 줬다.

sangwon21 committed:
feat: index 기반 슬라이더 기능 구현

why: 헤더 버튼을 index로 이용하여 슬라이더를 구현할 예정이다. 이를 위해,
기존의 버튼을 클릭하면 1칸씩 이동하던 방법을 index기반으로 이동하는
방식으로 교체하였다.

sangwon21 committed:
feat: 동그라미를 클릭시 슬라이더 이동 구현

why: 동그라미를 클릭하면 이동하는 이벤트를 구현했다. 인덱스 기반에서
거리 기반으로 이동하게끔 함수를 바꿨다. 거리 구하는 로직이 복잡해지는
만큼, 나머지 로직이 편해졌다.

sangwon21 committed:
chore: slider.html 파일 이름 변경

why: git page 등록 편의성을 위하여 slider.html을 index.html로 변경했다.

sangwon21 committed:
feat: 클릭시 카드 이동

why: 클릭시 그 카테고리의 첫번째 카드로 이동하는 기능을 구현했다. 이를
위해, 객체들 간 참조가 필요하다고 생각했다. 하지만 지금 구조에서 다른
이벤트 객체를 참조하지는 못한다. 그래서 구조를 바꾸는 고민도 했지만,
답은 의외로 click()이라는 함수를 통해 클릭을 흉내내는 방법으로 해결했다.
click을 하면 이벤트가 발생하니, click을 통하여 나머지 이벤트들을
해결했다.

sangwon21 committed:
chore: index.html 파일 이름 변경

why: slider.html 이름이 더 적절하다.

sangwon21 committed:
chore: update README

why: README에 동작하는 페이지 주소를 올렸다.

sangwon21 committed:
refact: 특정 상수 값 변수화

why: 알고리즘 작성 중에 필요한 값들이 기존에는 상수로 들어가 있었는데
변수화를 진행함으로써, 좀 더 알고리즘에 대한 이해를 높일 수 있게끔
작성했다.

sangwon21 committed:
refact: 공통된 부분 함수로 분리

why: 카드 이동 후, 나타난 카드에 효과를 주는 부분을 css클래스로
작성했다. 이 클래스를 줌으로써, 효과를 준다. 이 부분이 여러 함수에서
활용되기에, 독립적인 함수로 분리했다.

sangwon21 committed:
refact: 초기화 함수 부모 객체로 분리

why: event 객체에서 초기화 과정에 필요한 함수가 겹쳤다. 따라서 필요한 이
공통 함수를 부모 객체에 매서드로 주었다.

sangwon21 committed:
feat: 첫 화면 제공 시 필요한 CSS 클래스 추가

why: 첫 화면이 띄워졌을 때, 필요한 CSS 클래스들을 추가했다. 페이지에서
제공되는 효과들은 CSS 클래스에 기반하고 있는데, 첫 화면 렌더링 시 이
클래스들을 tag에 추가하지 못해, 첫 화면에서는 효과가 없었다. 이를
해결하기 위하여, 첫 번째 렌더링 때만 해당 CSS 클래스들을 추가했다.

sangwon21 committed:
chore: update README

why: git page 주소가 수정되어 README에 업데이트 했다.

sangwon21 committed:
refact: 공통된 부분 함수로 분리

why: 이벤트 콜백 함수에서 SliderEvent 클래스에 필요한 데이터를
초기화하는데 반복되는 코드가 있었다. 반복적으로 작성되는 코드를 분리하여
하나의 독립적인 함수로 만들었다.

crongro commented:
> * bind를 해서 this context를 강제하는 게 과연 올바른 함수 사용법인지 궁금합니다.

네 그렇게 하라고 만든함수라서 쓰면됩니다. arrow function을 사용하면 안해도 되는 경우가 있으니 확인해보세요.

crongro commented:
> * 이벤트 전달을 고려하여 코드를 작성하다 보니, 재활용성이 많이 떨어지는 함수를 작성했습니다. 특히, HTML 구조가 바꼈을 때, 유연하게 대처할 수 있는, 이벤트 전달을 고려한 함수를 작성하기 위해서는, 어떠한 영역이 필요할까요?

모든 함수를 재사용을 목적으로 구현하긴 어렵죠. 너무  재사용을 목적으로 생각하지는 않아도 될거고요.
가능한 부분에서만 지금의 생각을 유지하는 수준으로 구현해도 될듯해요.

crongro commented:
> * card 객체의 sortByCategory() 매서드 호출하는 곳을 살펴보면, for문 한 번으로 대체할 수 있습니다. 대신 sortByCategory()함수는 조금 더 가독성이 좋은 것 같습니다. 이런 상황에서 어떤 선택을 해야 하는지 궁금합니다.

어려운 질문이고, 제 생각이 정답은 아닙니다.  
성능이 중요한  ux가 아니라면, 가독성을 우선하고, 느리다면 변경합니다.ㅎ

crongro commented on button.js:
```diff
@@ -0,0 +1,15 @@
+                    (index + 1)}>${data}</button>`),
+            ""
+        );
+        return `${buttons}`;
```
reduce 잘 구현했어요.
buttonData보다는 buttons 라고 하는것도 좋을 듯.

crongro commented on button.js:
```diff
@@ -0,0 +1,15 @@
+
+    render() {
+        const buttons = this.buttonData.reduce(
+            (result, data, index) =>
```
index는 안쓸거면 생략 가능.

sangwon21 replied:
index를 쓰면 reduce를 쓰는 이유가 희석되나요?

crongro commented on card.js:
```diff
@@ -0,0 +1,23 @@
+class Card {
```
작은 컴포넌트로 나눈시도 좋아요.

crongro commented on card.js:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    sortByCategory(category) {
+        return this.cardData
```
filter-reduce 도 좋은방법이네요.

pipeline으로 함수 composition 하는 것도 나중에 시도를!

crongro commented on event-manager.js:
```diff
@@ -0,0 +1,18 @@
+class EventManager{
```
이것도 괜찮은 시도에요.

Event를 관장하는 녀석이 있어서 이곳에 몰아두는 것이죠. 
일종의 컨트롤러역할처럼 동작하는거죠.

crongro commented on event-manager.js:
```diff
@@ -0,0 +1,18 @@
+        $$(".dot").forEach(dot => {
+            this.sliderEvent.addEvent(dot, "click", this.sliderEvent.dotEventListener.bind(this.sliderEvent));
+        })
+        this.sliderEvent.addEvent($(".button2"), "click", this.sliderEvent.nextButtonListener.bind(this.sliderEvent));
```
bind는 이런상황에서 쓰라고 있는거니까~ 편안하게 생각하세요.
bind의 동작에 대해서 잘 이해를 하는게 오히려 중요하죠.

crongro commented on slider-event.js:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    giveSelectedClassToCurrentIndexNode() {
+        const selectedCard = $('.selected-card');
```
selectedCard가 반복적으로 나오면 this로 속성으로 추가해도 될 듯.

crongro commented on slider-event.js:
```diff
@@ -0,0 +1,139 @@
+        this.giveSelectedClassToCurrentIndexNode();
+    }
+
+    nextButtonListener(event) {
```
previousButtonListener 와 완전히  나누는 것도 좋은데, 
음 합칠 수 있는 방법도 있을거 같아요.

sangwon21 replied:
previousButtonListener와 nextButtonListener 각각 서로 다른 2개의 버튼에 이벤트를 달고 있습니다. 만약 합치게 되면 2개의 다른 버튼에 똑같은 함수를 Listener로 등록하게 됩니다. 비슷한 로직의 Listener들을 하나의 함수에서 처리하게 되면 복잡도가 증가하게 될 것 같습니다. 복잡도를 증가시키지 않으면서 함수를 하나로 뺄 수 있는 방법이 있을까요?

crongro commented on slider-event.js:
```diff
@@ -0,0 +1,139 @@
+        const firstParitialList = childNodes.slice(0, Math.abs(this.distance));
+        const firstSlide = childNodes[0];
+
+        if (this.isPrevious === false) {
```
조건문에서  === true, false 를 했는데..
엄격한 체크하는 거 아니면 if(this.isPrevious)  해도 사실 같은거라..

sangwon21 replied:
가독성 측면에서 명시적으로 if 조건문 안에 true와 false를 적었는데, true와 false를 생략하는 방법이 더 가독성이 좋을까요?

crongro commented on slider.js:
```diff
@@ -0,0 +1,15 @@
+class Slider {
```
껍데기역할만 하는 slider.js 가 단순해서 좋네요.

crongro commented on main.js:
```diff
@@ -0,0 +1,24 @@
+const addEvents = () => {
```
상속을 했으니, 장/단점을 스스로 정리해보세요.
상속이 anti-pattern만은 아닙니다!

sangwon21 replied:
댓글에 간단히 상속을 정리해봅니다.

## 장점과 단점

- 장점
    - 상위 클래스에서 필드 및 공통 메서드를 상속하여 사용할 수 있다. 따라서 하위클래스에서 소스코드의 양이 줄어들고 기능을 확장하기 용이하다.
- 단점
    - 상속 구조가 복잡해지면 상위클래스의 변화가 하위클래스에 주는 영향을 예측하기 힘들다. 적절하지 못한 상속을 사용하면 의도했던 것과 다르게 동작할 수 있다. 
- [출처](https://defacto-standard.tistory.com/157)

### 더 읽어볼 자료
- [5 Reasons to Prefer Composition over Inheritance in Java](https://javarevisited.blogspot.com/2013/06/why-favor-composition-over-inheritance-java-oops-design.html#ixzz6DpNlDGAM)

crongro commented on header-event.js:
```diff
@@ -0,0 +1,46 @@
+        this.initializeStatus();
+        
+        // headerTitle
+        if (event.target.classList.contains("header_title") === true) {
```
4가지조건문의 코드를 메서드로 분리해도 될 거 같네요.

crongro commented on header-event.js:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        // circles
+        if (event.target.classList.contains("circles") === true) {
```
event.target 이렇게 반복적으로 계속  찾지 말고, 변수로 담아두고 캐시해서 쓰시죠.
const {target} = event;

crongro commented on slider-event.js:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    getDistance(middle, target) {
+        if (middle + this.halfListLength > this.listLength) {
```
getDistance 메서드안에 조건문에  사용한 비교문들은 뭔지 잘 알기 어렵네요.
모두 함수로 분리하는 것도 방법이긴해요.

crongro reviewed (changes requested):
잘게 나누다 보니 좀 복잡한 관계로 보여지기도 하네요.

전체 클래스를 시각적으로 도형을 이용해서 그려보세요.
그리고 관계를 표시해보시고요.
역할이 정확하고, 흐름이 자연스러워보이는지 스스로 점검합니다.

사실 이정도 slide를 이렇게 잘게 나누는 건 과한 느낌이긴해요.
framework를 사용하면 그녀석이 담당해주는 부분이 많아서 위임할 수 있지만요.

어쨌든 좋은 시도고요! 

더 리팩토링할 게 있으면 시도해보세요!
PR은 request changes합니다.

sangwon21 commented:
![diagram](https://user-images.githubusercontent.com/47213425/74425482-19512d00-4e97-11ea-86dd-7032a6a11d4c.jpg)
- 말씀하신대로 적어보니 상속을 굳이 하지 않고, EventManager가 MyEvent 역할을 대신할 수 있을 것 같습니다.

sangwon21 committed:
refact: 반복되는 코드나 의미가 불명확한 부분을 함수로 분리

why: 반복적으로 사용되는 코드를 객체의 멤버 변수를 활용하여 없앴다.
특히, 반복되는 코드가 계속해서 dom 트리를 탐색하는 작업이기에, 성능에
도움이 된다. 또한, getDistance에서 알고리즘을 담당하는 부분에서, 의미가
불명확한 부분이 있기에, 작은 함수로 나누어 의미를 명확한 의미를
부여했다.

crongro reviewed (approved):
좋네요.
리팩토링은  설계를 통해서 점검해보는 게  좋은거 같더라고요.
그런 재미를 느껴보면서 계속 코드를 다듬어보세요.
(코드를 너무 사랑하지는 말기 ㅎ)
